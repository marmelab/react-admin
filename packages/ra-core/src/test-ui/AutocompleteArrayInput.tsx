import * as React from 'react';
import {
    FieldTitle,
    InputProps,
    isRequired,
    useChoicesContext,
    useInput,
} from '../';

export const AutocompleteArrayInput = (props: Partial<InputProps>) => {
    const { allChoices, source, setFilters, filterValues } =
        useChoicesContext();

    const { field, fieldState } = useInput({ source, ...props });

    return (
        <div>
            <div>
                <FieldTitle
                    label={props.label}
                    source={props.source}
                    resource={props.resource}
                    isRequired={isRequired(props.validate)}
                />
            </div>
            <input
                type="text"
                value={filterValues['q']}
                onChange={e =>
                    setFilters({ ...filterValues, q: e.target.value })
                }
            />
            <button type="button" onClick={() => field.onChange([])}>
                Clear value
            </button>
            <ul>
                {allChoices?.map(choice => (
                    <li key={choice.id}>
                        <label>
                            <input
                                type="checkbox"
                                value={choice.id}
                                onChange={event => {
                                    const newValue = event.target.checked
                                        ? [...field.value, choice.id]
                                        : field.value.filter(
                                              (v: any) => v !== choice.id
                                          );
                                    field.onChange(newValue);
                                }}
                                checked={field.value.includes(choice.id)}
                                aria-describedby={
                                    fieldState.error
                                        ? `error-${props.source}`
                                        : undefined
                                }
                            />
                            {choice.name}
                        </label>
                    </li>
                ))}
            </ul>
            {fieldState.error ? (
                <p id={`error-${props.source}`} style={{ color: 'red' }}>
                    {fieldState.error.message}
                </p>
            ) : null}
        </div>
    );
};
