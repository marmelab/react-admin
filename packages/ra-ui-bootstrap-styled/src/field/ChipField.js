import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import pure from 'recompose/pure';
import Chip from 'material-ui/Chip';
import { withStyles } from 'material-ui/styles';
import classnames from 'classnames';
import sanitizeRestProps from './sanitizeRestProps';

const styles = {
    chip: { margin: 4 },
};

export const ChipField = ({
    className,
    classes = {},
    source,
    record = {},
    ...rest
}) => {
    return (
        <Chip
            className={classnames(classes.chip, className)}
            label={get(record, source)}
            {...sanitizeRestProps(rest)}
        />
    );
};

ChipField.propTypes = {
    className: PropTypes.string,
    classes: PropTypes.object,
    elStyle: PropTypes.object,
    source: PropTypes.string.isRequired,
    record: PropTypes.object,
};

const PureChipField = withStyles(styles)(pure(ChipField));

export default PureChipField;
