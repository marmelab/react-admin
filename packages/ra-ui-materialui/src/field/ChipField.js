import React from "react";
import PropTypes from "prop-types";
import get from "lodash/get";
import pure from "recompose/pure";
import Chip from "@material-ui/core/Chip";
import { withStyles } from "@material-ui/core/styles";
import classnames from "classnames";
import sanitizeRestProps from "./sanitizeRestProps";

const styles = {
  chip: { margin: 4 }
};

export const ChipField = ({
  className,
  classes = {},
  source,
  record = {},
  ...rest
}) => {
  return (
    <Chip
      className={classnames(classes.chip, className)}
      label={get(record, source)}
      {...sanitizeRestProps(rest)}
    />
  );
};

ChipField.propTypes = {
  className: PropTypes.string,
  classes: PropTypes.object,
  elStyle: PropTypes.object,
  sortBy: PropTypes.string,
  source: PropTypes.string.isRequired,
  record: PropTypes.object
};

// wat? TypeScript looses the displayName if we don't set it explicitly
ChipField.displayName = "ChipField";

const PureChipField = withStyles(styles)(pure(ChipField));

export default PureChipField;
