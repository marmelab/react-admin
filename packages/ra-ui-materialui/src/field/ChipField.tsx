import React, { FunctionComponent } from 'react';
import compose from 'recompose/compose';
import get from 'lodash/get';
import pure from 'recompose/pure';
import Chip, { ChipProps } from '@material-ui/core/Chip';
import { makeStyles } from '@material-ui/core/styles';
import classnames from 'classnames';

import sanitizeRestProps from './sanitizeRestProps';
import { FieldProps, InjectedFieldProps, fieldPropTypes } from './types';

const useStyles = makeStyles(
    {
        chip: { margin: 4 },
    },
    { name: 'RaChipField' }
);

export const ChipField: FunctionComponent<
    FieldProps & InjectedFieldProps & ChipProps
> = ({ className, classes: classesOverride, source, record = {}, ...rest }) => {
    const classes = useStyles({ classes: classesOverride });

    return (
        <Chip
            className={classnames(classes.chip, className)}
            label={get(record, source)}
            {...sanitizeRestProps(rest)}
        />
    );
};

const EnhancedChipField = compose<
    FieldProps & InjectedFieldProps & ChipProps,
    FieldProps & ChipProps
>(pure)(ChipField);

EnhancedChipField.defaultProps = {
    addLabel: true,
};

EnhancedChipField.propTypes = {
    ...ChipField.propTypes,
    ...fieldPropTypes,
};

EnhancedChipField.displayName = 'EnhancedChipField';

export default EnhancedChipField;
