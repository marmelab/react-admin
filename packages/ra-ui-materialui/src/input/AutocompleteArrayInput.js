import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash.get';
import compose from 'recompose/compose';
import { withStyles } from 'material-ui/styles';
import { AutocompleteInput } from './AutocompleteInput';
import { addField, translate } from 'ra-core';

const styles = theme => ({
    chips: {
        display: 'flex',
    },
    container: {
        flexGrow: 1,
        position: 'relative',
    },
    root: {},
    suggestionsContainerOpen: {
        position: 'absolute',
        marginBottom: theme.spacing.unit * 3,
        zIndex: 2,
    },
    suggestion: {
        display: 'block',
        fontFamily: theme.typography.fontFamily,
    },
    suggestionText: { fontWeight: 300 },
    highlightedSuggestionText: { fontWeight: 500 },
    suggestionsList: {
        margin: 0,
        padding: 0,
        listStyleType: 'none',
    },
});

class AutocompleteArrayInput extends React.Component {
    static propTypes = {
        choices: PropTypes.arrayOf(PropTypes.object),
        children: PropTypes.element,
        classes: PropTypes.object,
        input: PropTypes.shape({
            value: PropTypes.array,
        }),
        optionText: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
        optionValue: PropTypes.string,
        onChipAdded: PropTypes.func,
        onChipDeleted: PropTypes.func,
        onChipClicked: PropTypes.func,
        selectedItems: PropTypes.arrayOf(PropTypes.object),
        setFilter: PropTypes.func,
        translate: PropTypes.func,
        translateChoice: PropTypes.bool,
    };
    static defaultProps = {
        addField: true,
        choices: [],
        classes: {},
        optionValue: 'id',
        optionText: 'name',
        fullWidth: true,
    };

    constructor(...args) {
        super(...args);
        this.state = {
            chipData: [],
            choices: [],
            input: {
                value: [],
                ...this.props.input,
                onChange: this.handleChipAdd,
                onBlur: this.handleBlur,
            },
        };
    }

    componentWillMount() {
        this.updateItems();
    }

    componentWillReceiveProps(nextProps) {
        const { input, choices } = this.props;
        if (
            nextProps.input.value !== input.value ||
            nextProps.choices !== choices
        ) {
            this.updateItems(nextProps);
        }
    }

    updateItems = ({ choices, input } = this.props) => {
        const inputValue = Array.isArray(input.value) ? input.value : [];

        const choicesGroupedByType = choices.reduce(
            (acc, choice) => {
                const selectedIndex = inputValue.indexOf(
                    this.getOptionValue(choice)
                );
                const target = selectedIndex !== -1 ? 'chipData' : 'choices';

                acc[target][
                    target === 'chipData' ? selectedIndex : acc[target].length
                ] = choice;

                return acc;
            },
            {
                chipData: [],
                choices: [],
            }
        );
        this.setState({
            chipData: this.props.selectedItems || choicesGroupedByType.chipData,
            choices: choicesGroupedByType.choices,
        });
    };

    getOptionValue = option => get(option, this.props.optionValue);

    handleChipAdd = data => {
        const { onChipAdded, input } = this.props;
        onChipAdded && onChipAdded(data);

        input.onChange(input.value ? input.value.concat(data) : [data]);

        this.setState(({ input }) => ({
            input: {
                ...input,
                value: `RA_FAKE_ID_${Date.now()}`, // This ensures Autocomplete will rerender and it doesn't find a matching choice.
            },
        }));
    };

    handleChipDelete = data => () => {
        const { onChipDeleted, input } = this.props;
        onChipDeleted && onChipDeleted(data);

        const chipValue = this.getOptionValue(data);
        input.onChange(input.value.filter(it => it !== chipValue));
    };

    handleChipClick = data => {
        const { onChipClicked } = this.props;
        onChipClicked && onChipClicked(data);
    };

    handleBlur = () => {
        const { input } = this.props;
        input.onBlur(input.value);
    };

    handleFilter = text => {
        const { setFilter } = this.props;
        setFilter && setFilter(text);
    };

    render() {
        const { children, classes, ...props } = this.props;
        const { chipData } = this.state;
        return (
            <AutocompleteInput
                classes={classes}
                {...props}
                choices={this.state.choices}
                input={this.state.input}
                setFilter={this.handleFilter}
                options={{
                    InputProps: {
                        startAdornment: (
                            <div className={classes.chips}>
                                {chipData.map(chip =>
                                    React.cloneElement(children, {
                                        key: this.getOptionValue(chip),
                                        record: chip,
                                        onClick: this.handleChipClick,
                                        onDelete: this.handleChipDelete(chip),
                                    })
                                )}
                            </div>
                        ),
                    },
                }}
            />
        );
    }
}

const enhance = compose(addField, translate, withStyles(styles));
export { AutocompleteArrayInput };
export default enhance(AutocompleteArrayInput);
