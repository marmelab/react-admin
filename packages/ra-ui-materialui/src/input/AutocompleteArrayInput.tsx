import * as React from 'react';
import { RaRecord } from 'ra-core';
import { AutocompleteInput, AutocompleteInputProps } from './AutocompleteInput';

/**
 * An Input component for an autocomplete field, using an array of objects for the options
 *
 * Pass possible options as an array of objects in the 'choices' attribute.
 *
 * By default, the options are built from:
 *  - the 'id' property as the option value,
 *  - the 'name' property as the option text
 * @example
 * const choices = [
 *    { id: 'M', name: 'Male' },
 *    { id: 'F', name: 'Female' },
 * ];
 * <AutocompleteArrayInput source="gender" choices={choices} />
 *
 * You can also customize the properties to use for the option name and value,
 * thanks to the 'optionText' and 'optionValue' attributes.
 * @example
 * const choices = [
 *    { _id: 123, full_name: 'Leo Tolstoi', sex: 'M' },
 *    { _id: 456, full_name: 'Jane Austen', sex: 'F' },
 * ];
 * <AutocompleteArrayInput source="author_id" choices={choices} optionText="full_name" optionValue="_id" />
 *
 * `optionText` also accepts a function, so you can shape the option text at will:
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 * <AutocompleteArrayInput source="author_id" choices={choices} optionText={optionRenderer} />
 *
 * `optionText` also accepts a React Element, that can access
 * the related choice through the `useRecordContext` hook. You can use Field components there.
 * Note that you must also specify the `matchSuggestion` and `inputText` props
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const matchSuggestion = (filterValue, choice) => choice.first_name.match(filterValue) || choice.last_name.match(filterValue)
 * const inputText = (record) => `${record.fullName} (${record.language})`;
 *
 * const FullNameField = () => {
 *     const record = useRecordContext();
 *     return <span>{record.first_name} {record.last_name}</span>;
 * }
 *
 * <AutocompleteArrayInput source="gender" choices={choices} optionText={<FullNameField />} matchSuggestion={matchSuggestion} />
 *
 * The choices are translated by default, so you can use translation identifiers as choices:
 * @example
 * const choices = [
 *    { id: 'M', name: 'myroot.gender.male' },
 *    { id: 'F', name: 'myroot.gender.female' },
 * ];
 *
 * However, in some cases (e.g. inside a `<ReferenceInput>`), you may not want
 * the choice to be translated. In that case, set the `translateChoice` prop to false.
 * @example
 * <AutocompleteArrayInput source="gender" choices={choices} translateChoice={false}/>
 */

export const AutocompleteArrayInput = <
    OptionType extends RaRecord = RaRecord,
    DisableClearable extends boolean | undefined = boolean | undefined,
    SupportCreate extends boolean | undefined = false
>({
    defaultValue,
    ...props
}: AutocompleteArrayInputProps<
    OptionType,
    DisableClearable,
    SupportCreate
>) => (
    <AutocompleteInput<OptionType, true, DisableClearable, SupportCreate>
        {...props}
        multiple
        defaultValue={defaultValue ?? []}
    />
);

export type AutocompleteArrayInputProps<
    OptionType extends any = RaRecord,
    DisableClearable extends boolean | undefined = false,
    SupportCreate extends boolean | undefined = false
> = Omit<
    AutocompleteInputProps<OptionType, true, DisableClearable, SupportCreate>,
    'defaultValue'
> & {
    defaultValue?: any[];
};
