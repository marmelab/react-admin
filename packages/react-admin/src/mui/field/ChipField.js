import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash.get';
import pure from 'recompose/pure';
import Chip from 'material-ui/Chip';
import { withStyles } from 'material-ui/styles';
import classnames from 'classnames';

const styles = {
    chip: { margin: 4 },
};

const ChipField = ({ className, classes, source, record = {} }) => (
    <Chip
        className={classnames(classes.chip, className)}
        label={get(record, source)}
    />
);

ChipField.propTypes = {
    addLabel: PropTypes.bool,
    className: PropTypes.string,
    classes: PropTypes.object,
    elStyle: PropTypes.object,
    label: PropTypes.string,
    source: PropTypes.string.isRequired,
    record: PropTypes.object,
};

const PureChipField = withStyles(styles)(pure(ChipField));

PureChipField.defaultProps = {
    addLabel: true,
};

export default PureChipField;
