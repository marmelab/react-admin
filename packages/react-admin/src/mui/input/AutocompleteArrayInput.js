import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash.get';
import classnames from 'classnames';
import compose from 'recompose/compose';
import { withStyles } from 'material-ui/styles';
import AutocompleteInput from './AutocompleteInput';
import translate from '../../i18n/translate';

const styles = () => ({
    root: {},
    chips: {
        display: 'flex',
    },
});

class AutocompleteArrayInput extends React.Component {
    static propTypes = {
        choices: PropTypes.arrayOf(PropTypes.object),
        classes: PropTypes.object,
        optionText: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
        optionValue: PropTypes.string,
        onChipAdded: PropTypes.func,
        onChipDeleted: PropTypes.func,
        onChipClicked: PropTypes.func,
        selectedItems: PropTypes.arrayOf(PropTypes.object),
        setFilter: PropTypes.func,
        translate: PropTypes.func,
        translateChoice: PropTypes.bool,
    };
    static defaultProps = {
        addField: true,
        optionValue: 'id',
        optionText: 'name',
        fullWidth: true,
    };

    constructor(...args) {
        super(...args);
        this.state = {
            chipData: [],
            choices: [],
            input: {
                ...this.props.input,
                value: '',
                onChange: this.handleChipAdd,
                onBlur: this.handleBlur,
            },
        };
    }
    componentWillMount() {
        this.updateItems();
    }
    componentWillReceiveProps(nextProps) {
        const { input, choices } = this.props;
        if (
            nextProps.input.value !== input.value ||
            nextProps.choices !== choices
        ) {
            this.updateItems(nextProps);
        }
    }

    updateItems = ({ choices, input } = this.props) => {
        const inputValue = Array.isArray(input.value) ? input.value : [];

        const choicesGroupedByType = choices.reduce(
            (acc, choice) => {
                const selectedIndex = inputValue.indexOf(
                    this.getOptionValue(choice)
                );
                const target = selectedIndex !== -1 ? 'chipData' : 'choices';

                acc[target][
                    target === 'chipData' ? selectedIndex : acc[target].length
                ] = choice;

                return acc;
            },
            {
                chipData: [],
                choices: [],
            }
        );
        this.setState({
            chipData: this.props.selectedItems || choicesGroupedByType.chipData,
            choices: choicesGroupedByType.choices,
        });
    };

    getOptionValue = option => get(option, this.props.optionValue);

    handleChipAdd = data => {
        const { onChipAdded, input } = this.props;
        onChipAdded && onChipAdded(data);

        input.onChange(input.value ? input.value.concat(data) : [data]);

        this.setState(({ input }) => ({
            input: {
                ...input,
                value: `RA_FAKE_ID_${Date.now()}`, // This ensures Autocomplete will rerender and it doesn't find a matching choice.
            },
        }));
    };

    handleChipDelete = data => () => {
        const { onChipDeleted, input } = this.props;
        onChipDeleted && onChipDeleted(data);

        const chipValue = this.getOptionValue(data);
        input.onChange(input.value.filter(it => it !== chipValue));
    };

    handleChipClick = data => {
        const { onChipClicked } = this.props;
        onChipClicked && onChipClicked(data);
    };

    handleBlur = () => {
        const { input } = this.props;
        input.onBlur(input.value);
    };

    handleFilter = text => {
        const { setFilter } = this.props;
        setFilter && setFilter(text);
    };

    render() {
        const { children, classes, className, ...props } = this.props;
        const { chipData } = this.state;
        return (
            <AutocompleteInput
                className={classnames(classes.root, className)}
                {...props}
                choices={this.state.choices}
                input={this.state.input}
                setFilter={this.handleFilter}
                options={{
                    InputProps: {
                        startAdornment: (
                            <div className={classes.chips}>
                                {chipData.map(chip =>
                                    React.cloneElement(children, {
                                        key: this.getOptionValue(chip),
                                        record: chip,
                                        onClick: this.handleChipClick,
                                        onDelete: this.handleChipDelete(chip),
                                    })
                                )}
                            </div>
                        ),
                    },
                }}
            />
        );
    }
}

const enhance = compose(translate, withStyles(styles));
export { AutocompleteArrayInput };
export default enhance(AutocompleteArrayInput);
