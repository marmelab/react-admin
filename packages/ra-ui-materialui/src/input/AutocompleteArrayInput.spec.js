import React from 'react';
import assert from 'assert';
import { shallow } from 'enzyme';
import { AutocompleteArrayInput } from './AutocompleteArrayInput';
import ChipField from '../field/ChipField';

describe('<AutocompleteArrayInput />', () => {
    const defaultProps = {
        source: 'foo',
        meta: {},
        input: {},
        translate: x => x,
    };

    it('should use a AutocompleteInput', () => {
        const wrapper = shallow(
            <AutocompleteArrayInput
                {...defaultProps}
                input={{ value: 1 }}
                choices={[{ id: 1, name: 'hello' }]}
            />
        );
        expect(wrapper.find('AutocompleteInput')).toHaveLength(1);
    });

    it('should have an empty input suggestion text on initial value', () => {
        const wrapper = shallow(
            <AutocompleteArrayInput
                {...defaultProps}
                input={{ value: ['programming', 'lifestyle'] }}
            />
        );
        assert.deepEqual(
            wrapper
                .find('AutocompleteInput')
                .dive()
                .prop('inputProps').value,
            ''
        );
        expect(wrapper.state('input').value).toEqual([
            'programming',
            'lifestyle',
        ]);
    });

    it('should render choices', () => {
        const wrapper = shallow(
            <AutocompleteArrayInput
                {...defaultProps}
                choices={[
                    { id: 'programming', name: 'Programming' },
                    { id: 'lifestyle', name: 'Lifestyle' },
                    { id: 'photography', name: 'Photography' },
                ]}
            >
                <ChipField source="name" />
            </AutocompleteArrayInput>
        );
        expect(wrapper).toMatchSnapshot();
    });
    it('should render chips of input value and choices for the rest', () => {
        const wrapper = shallow(
            <AutocompleteArrayInput
                {...defaultProps}
                input={{ value: ['programming'] }}
                choices={[
                    { id: 'programming', name: 'Programming' },
                    { id: 'lifestyle', name: 'Lifestyle' },
                    { id: 'photography', name: 'Photography' },
                ]}
            >
                <ChipField source="name" />
            </AutocompleteArrayInput>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
