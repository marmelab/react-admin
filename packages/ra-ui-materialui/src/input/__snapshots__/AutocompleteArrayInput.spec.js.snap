// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteArrayInput /> should render chips of input value and choices for the rest 1`] = `
<AutocompleteInput
  addField={true}
  choices={
    Array [
      Object {
        "id": "lifestyle",
        "name": "Lifestyle",
      },
      Object {
        "id": "photography",
        "name": "Photography",
      },
    ]
  }
  classes={Object {}}
  fullWidth={true}
  input={
    Object {
      "onBlur": [Function],
      "onChange": [Function],
      "value": Array [
        "programming",
      ],
    }
  }
  inputValueMatcher={[Function]}
  limitChoicesToValue={false}
  meta={Object {}}
  optionText="name"
  optionValue="id"
  options={
    Object {
      "InputProps": Object {
        "startAdornment": <div
          className={undefined}
    >
          <WithStyles(pure(ChipField))
                onClick={[Function]}
                onDelete={[Function]}
                record={
                      Object {
                            "id": "programming",
                            "name": "Programming",
                          }
                }
                source="name"
          />
    </div>,
      },
    }
  }
  setFilter={[Function]}
  source="foo"
  translate={[Function]}
  translateChoice={true}
/>
`;

exports[`<AutocompleteArrayInput /> should render choices 1`] = `
<AutocompleteInput
  addField={true}
  choices={
    Array [
      Object {
        "id": "programming",
        "name": "Programming",
      },
      Object {
        "id": "lifestyle",
        "name": "Lifestyle",
      },
      Object {
        "id": "photography",
        "name": "Photography",
      },
    ]
  }
  classes={Object {}}
  fullWidth={true}
  input={
    Object {
      "onBlur": [Function],
      "onChange": [Function],
      "value": Array [],
    }
  }
  inputValueMatcher={[Function]}
  limitChoicesToValue={false}
  meta={Object {}}
  optionText="name"
  optionValue="id"
  options={
    Object {
      "InputProps": Object {
        "startAdornment": <div
          className={undefined}
    >
          
    </div>,
      },
    }
  }
  setFilter={[Function]}
  source="foo"
  translate={[Function]}
  translateChoice={true}
/>
`;
